transport/disorder/stream_dialer.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package disorder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+var defaultTTL = 64
+
+type disorderDialer struct {
+	dialer     transport.StreamDialer
+	splitPoint int64
+}
+
+var _ transport.StreamDialer = (*disorderDialer)(nil)
+
+// NewStreamDialer creates a [transport.StreamDialer] that splits the outgoing stream after writing "prefixBytes" bytes
+// using [SplitWriter].
+func NewStreamDialer(dialer transport.StreamDialer, prefixBytes int64) (transport.StreamDialer, error) {
+	if dialer == nil {
+		return nil, errors.New("argument dialer must not be nil")
+	}
+	return &disorderDialer{dialer: dialer, splitPoint: prefixBytes}, nil
+}
+
+// DialStream implements [transport.StreamDialer].DialStream.
+func (d *disorderDialer) DialStream(ctx context.Context, remoteAddr string) (transport.StreamConn, error) {
+	innerConn, err := d.dialer.DialStream(ctx, remoteAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// this strategy only works when we set TCP as a strategy
+	tcpConn, ok := innerConn.(*net.TCPConn)
+	if !ok {
+		return nil, fmt.Errorf("split strategy only works with direct TCP connections")
+	}
+
+	err = tcpConn.SetNoDelay(true)
+	if err != nil {
+		return nil, fmt.Errorf("setting tcp NO_DELAY failed: %w", err)
+	}
+
+	file, err := tcpConn.File()
```

Comment:
Use [syscall.RawConn.Control](https://pkg.go.dev/syscall#RawConn) instead

Reply:
changed to use x/net